File: Dished/View Controllers/DANotificationSettingsViewController.m
//
//  DANotificationSettingsViewController.m
//  Dished
//
//  Created by Ryan Khalili on 10/12/14.
//  Copyright (c) 2014 Dished. All rights reserved.
//

#import "DANotificationSettingsViewController.h"
#import "DAUserManager.h"


@interface DANotificationSettingsViewController()

@end


@implementation DANotificationSettingsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self populateSetting];
}

- (void)populateSetting
{
    if( [self.notificationType isEqualToString:@"yum"] )
    {
        self.title = @"YUM Notifications";
        [self selectRowWithPushSetting:[DAUserManager sharedManager].receivesYumPushNotifications];
    }
    else if( [self.notificationType isEqualToString:@"comment"] )
    {
        self.title = @"Comment Notifications";
        [self selectRowWithPushSetting:[DAUserManager sharedManager].receivesCommentPushNotifications];
    }
    else if( [self.notificationType isEqualToString:@"review"] )
    {
        self.title = @"New Review Notifications";
        [self selectRowWithPushSetting:[DAUserManager sharedManager].receivesReviewPushNotifications];
    }
}

- (void)selectRowWithPushSetting:(ePushSetting)pushSetting
{
    switch( pushSetting )
    {
        case ePushSettingOff:      self.offCell.accessoryType      = UITableViewCellAccessoryCheckmark; break;
        case ePushSettingFollowed: self.followCell.accessoryType   = UITableViewCellAccessoryCheckmark; break;
        case ePushSettingEveryone: self.everyoneCell.accessoryType = UITableViewCellAccessoryCheckmark; break;
    }
}

- (void)deselectAllRows
{
    self.offCell.accessoryType      = UITableViewCellAccessoryNone;
    self.followCell.accessoryType   = UITableViewCellAccessoryNone;
    self.everyoneCell.accessoryType = UITableViewCellAccessoryNone;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    switch( indexPath.row )
    {
        case 0: [self selectedPushSetting:ePushSettingOff];      break;
        case 1: [self selectedPushSetting:ePushSettingFollowed]; break;
        case 2: [self selectedPushSetting:ePushSettingEveryone]; break;
    }
}
                 
- (void)selectedPushSetting:(ePushSetting)pushSetting
{
    [self deselectAllRows];
    
    [self selectRowWithPushSetting:pushSetting];
    
    if( [self.notificationType isEqualToString:@"yum"] )
    {
        [[DAUserManager sharedManager] setYumPushNotificationSetting:pushSetting completion:^( BOOL success )
        {
            [self populateSetting];
        }];
    }
    else if( [self.notificationType isEqualToString:@"comment"] )
    {
        [[DAUserManager sharedManager] setCommentPushNotificationSetting:pushSetting completion:^( BOOL success )
        {
            [self populateSetting];
        }];
    }
    else if( [self.notificationType isEqualToString:@"review"] )
    {
        [[DAUserManager sharedManager] setReviewPushNotificationSetting:pushSetting completion:^( BOOL success )
        {
            [self populateSetting];
        }];
    }
}

@end

File: Dished/View Controllers/DANotificationSettingsViewController.swift
//
//  DANotificationSettingsViewController2.swift
//  Dished
//
//  Created by Ryan Khalili on 5/7/15.
//  Copyright (c) 2015 Dished. All rights reserved.
//

import UIKit

enum DANotificationSettingType {
    case Yum
    case Comment
    
    func name() -> String {
        switch(self) {
            case .Yum: return "YUM Notifications"
            case .Comment: return "Comment Notifications"
        }
    }
}

class DANotificationSettingsViewController: DAViewController, UITableViewDelegate, UITableViewDataSource {

    var notificationSettingsView = DANotificationSettingsView()
    let notificationSettingsDataSource = DANotificationSettingsDataSource()
    var notificationSetting = DANotificationSettingType.Yum
    
    private let cellIdentifier = "notificationSettingCell"
    
    init(notificationSetting: DANotificationSettingType) {
        super.init(nibName: nil, bundle: nil)
        
        navigationItem.title = notificationSetting.name()
        self.notificationSetting = notificationSetting
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        notificationSettingsView.tableView.delegate = self
        notificationSettingsView.tableView.dataSource = self
        notificationSettingsView.tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    }

    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 3
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) as! UITableViewCell
        
        let pushSetting = notificationSettingsDataSource.pushSettings[indexPath.row]
        cell.textLabel?.text = pushSetting.name()
        cell.textLabel?.font = DAConstants.primaryFontWithSize(17.0)
        
        let checkmark = UITableViewCellAccessoryType.Checkmark
        let isSelected = pushSetting == notificationSettingsDataSource.currentPushSettingForNotificationSettingType(notificationSetting)
        cell.accessoryType = isSelected ? checkmark : UITableViewCellAccessoryType.None
        
        return cell
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let dataSource = notificationSettingsDataSource
        let pushSetting = dataSource.pushSettings[indexPath.row]
        
        switch(notificationSetting) {
            case .Yum: dataSource.userManager.yumPushSetting = pushSetting
            case .Comment: dataSource.userManager.commentPushSetting = pushSetting
        }
        
        tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: UITableViewRowAnimation.Automatic)
    }
    
    override func loadView() {
        view = notificationSettingsView
    }
}